### PR TITLE
Contract tx limiter testing module

### PR DESCRIPTION
**Pull request by suhasagg**
- [ ] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [ ] All IP is original and not copied from another source
- [ ] I assign all copyright to Loom Network for the code in the pull request
1)Adds e2e test cases
2)loom.yaml and genesis.json for contract Tx Limiter
3)CLI for contract Tx Limiter
4)Block range and Max Txs Input Data Validation

### CODE REVIEW

File: cmd/loom/userdeployerwhitelist/cmd.go [hunk 1]
```diff
@@ -263,16 +264,42 @@ func setTierInfoCmd() *cobra.Command {
...
+				BlockRange: uint64(blockrange),
+				MaxTxs:     uint64(maxtxs),
+				TierID:     udwtypes.TierID(tierID),
 			}
 			return cli.CallContractWithFlags(&flags, dwContractName, "SetTierInfo", req, nil)
 		}}
 
 	cmd.Flags().StringVarP(&inputFee, "fee", "f", "", "Tier fee")
 	cmd.Flags().StringVarP(&tierName, "name", "n", "", "Tier name")
+	cmd.Flags().StringVarP(&blockRange, "blockRange", "b", "", "Block range")
```
Comment (enlight) [inline cmd/loom/userdeployerwhitelist/cmd.go]: `blockRange` -> `block-range`

File: cmd/loom/userdeployerwhitelist/cmd.go [hunk 2]
```diff
@@ -263,16 +264,42 @@ func setTierInfoCmd() *cobra.Command {
...
+	cmd.Flags().StringVarP(&maxTxs, "maxTxs", "t", "", "Max transactions")
```
Comment (enlight) [inline cmd/loom/userdeployerwhitelist/cmd.go]: `maxTxs` -> `max-txs`

File: cmd/loom/userdeployerwhitelist/cmd.go [hunk 3]
```diff
@@ -263,16 +264,42 @@ func setTierInfoCmd() *cobra.Command {
 			if len(tierName) == 0 {
 				tierName = getTierInfoResp.Tier.Name
 			}
+			if len(blockRange) == 0 {
+				blockrange = int64(getTierInfoResp.Tier.BlockRange)
+			} else {
+				blockrange, err = strconv.ParseInt(blockRange, 10, 64)
+				if err != nil {
+					return errors.Wrapf(err, "blockRange %s does not parse as integer", blockRange)
```
Comment (enlight) [inline cmd/loom/userdeployerwhitelist/cmd.go]: `blockRange` -> `block range`

File: cmd/loom/userdeployerwhitelist/cmd.go [hunk 4]
```diff
@@ -263,16 +264,42 @@ func setTierInfoCmd() *cobra.Command {
...
+				}
+				if blockrange == 0 {
+					return fmt.Errorf("blockrange must be greater than zero")
```
Comment (enlight) [inline cmd/loom/userdeployerwhitelist/cmd.go]: `blockrange` -> `block range`

File: cmd/loom/userdeployerwhitelist/cmd.go [hunk 5]
```diff
@@ -263,16 +264,42 @@ func setTierInfoCmd() *cobra.Command {
...
+				}
+			}
+			if len(maxTxs) == 0 {
+				maxtxs = int64(getTierInfoResp.Tier.MaxTxs)
+			} else {
+				maxtxs, err = strconv.ParseInt(maxTxs, 10, 64)
+				if err != nil {
+					return errors.Wrapf(err, "maxTxs %s does not parse as integer", maxTxs)
```
Comment (enlight) [inline cmd/loom/userdeployerwhitelist/cmd.go]: `maxTxs` -> `max-txs`

File: cmd/loom/userdeployerwhitelist/cmd.go [hunk 6]
```diff
@@ -263,16 +264,42 @@ func setTierInfoCmd() *cobra.Command {
...
+				}
+				if maxtxs == 0 {
+					return fmt.Errorf("maxtxs must be greater than zero")
```
Comment (enlight) [inline cmd/loom/userdeployerwhitelist/cmd.go]: `maxtxs` -> `max-txs`

File: e2e/userdeployerwhitelist.genesis.json [hunk 7]
```diff
@@ -1,105 +1,112 @@
 {
-    "contracts": [
-        {
-            "vm": "plugin",
-            "format": "plugin",
-            "name": "coin",
-            "location": "coin:1.0.0",
-            "init": {
-            }
+  "contracts":[
```
Comment (enlight) [inline e2e/userdeployerwhitelist.genesis.json]: Format this properly

File: cmd/loom/userdeployerwhitelist/cmd.go (hunk 1 above)
Comment (suhasagg) [reply cmd/loom/userdeployerwhitelist/cmd.go]: Done.

File: cmd/loom/userdeployerwhitelist/cmd.go (hunk 2 above)
Comment (suhasagg) [reply cmd/loom/userdeployerwhitelist/cmd.go]: Done.

File: cmd/loom/userdeployerwhitelist/cmd.go (hunk 3 above)
Comment (suhasagg) [reply cmd/loom/userdeployerwhitelist/cmd.go]: Done.

File: cmd/loom/userdeployerwhitelist/cmd.go (hunk 4 above)
Comment (suhasagg) [reply cmd/loom/userdeployerwhitelist/cmd.go]: Done.

File: cmd/loom/userdeployerwhitelist/cmd.go (hunk 5 above)
Comment (suhasagg) [reply cmd/loom/userdeployerwhitelist/cmd.go]: Done.

File: cmd/loom/userdeployerwhitelist/cmd.go (hunk 6 above)
Comment (suhasagg) [reply cmd/loom/userdeployerwhitelist/cmd.go]: Done.

File: e2e/userdeployerwhitelist.genesis.json (hunk 7 above)
Comment (suhasagg) [reply e2e/userdeployerwhitelist.genesis.json]: Done.

File: e2e/userdeployerwhitelist.toml [hunk 8]
```diff
@@ -1,77 +1,370 @@
...
 
 [[TestCases]]
-  RunCmd = "{{ $.LoomPath }} dev add-deployer --tier 0 {{index $.AccountAddressList 1}} -k {{index $.AccountPrivKeyPathList 0}}"
-  Condition = "contains"
-  Expected = ['']
+Condition = "contains"
+Expected = ['']
+RunCmd = """
+{{ $.LoomPath }} dev add-deployer --tier 0 {{index $.AccountAddressList 1}} -k {{index $.AccountPrivKeyPathList \
+0}}"""
```
[the remark this answers is not in the document]
Comment (suhasagg) [reply e2e/userdeployerwhitelist.toml]: Actually, I was using autoformatting tool :), It generated this toml file.

**Comment by suhasagg** [comment]
``` 
                        if blockRange == 0 {
				blockRange = getTierInfoResp.Tier.BlockRange
			}
			if blockRange == 0 {
				return fmt.Errorf("block range must be greater than zero")
			}
			if maxTxs == 0 {
				maxTxs = getTierInfoResp.Tier.MaxTxs
			}
			if maxTxs == 0 {
				return fmt.Errorf("max-txs must be greater than zero")
			}
		       ```
		       
		       @enlight this change is causing test to give incorrect output 
		      ==> expect output to contain 'block range must be greater than zero' got 'Error: [UserDeployerWhitelist] not authorized

		      
		      Command ran  -
		      run: /media/suhas/BEC6F1E3C6F19C39/go/whitelistcontract/src/github.com/loomnetwork/loomchain/loom dev set-tier 0 --fee 10 --name Tier2 --block-range 0 --max-txs 2 -k /media/suhas/BEC6F1E3C6F19C39/go/whitelistcontract/src/github.com/loomnetwork/loomchain/e2e/test-data/userdeployerwhitelist/privkey-0 -u http://127.0.0.1:35567

File: e2e/userdeployerwhitelist.toml [hunk 9]
```diff
@@ -1,77 +1,370 @@
 [[TestCases]]
-  Dir = ""
-  RunCmd = "{{ $.LoomPath }} coin balance {{index $.AccountAddressList 0}}"
-  Condition = "contains"
-  Expected = ["100000000000000000000"]
+Condition = "contains"
+Dir = ""
+Expected = ["100000000000000000000"]
+RunCmd = "{{ $.LoomPath }} coin balance {{index $.AccountAddressList 0}}"
```
[the remark this answers is not in the document]
Comment (suhasagg) [reply e2e/userdeployerwhitelist.toml]: Done.

File: e2e/userdeployerwhitelist.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (suhasagg) [reply e2e/userdeployerwhitelist.toml]: Done.

**Comment by suhasagg** [comment]
Please note test cases don't verify Error of transaction limit reached as block range is uniform and consecutive block height can fall in next interval as well making results slight unpredictable.

**Comment by suhasagg** [comment]
For example - if you consider block range of 5 with maximum 2 transactions.

**Comment by enlight** [comment]
> Please note test cases don't verify Error of transaction limit reached as block range is uniform and consecutive block height can fall in next interval as well making results slight unpredictable.

So what does this test verify?

**Comment by suhasagg** [comment]
This test can be used for generating data, how Tx capping is working with changing blockheight for multiple contracts which can then be verified.

**Pull request merged**